### ISSUE TITLE
Dilated convolution does not preserve tensor shape

### ISSUE THREAD

**Issue by haraldurt** [tensorflow/tensorflow#11954]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 14.04.5 LTS
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
('unknown', '1.3.0-rc0')
- **Python version**: 
2.7.12

- **Exact command to reproduce**:
```
input_tensor = tf.placeholder(tf.float32, (10, None, 256, 3))

dilated = tf.nn.convolution(input_tensor,
                            tf.zeros((3, 1, 3, 16)),
                            dilation_rate=[2, 1],
                            padding='SAME')

print(dilated.get_shape()) # Displays: [10, ?, ?, 16], expected [10, ?, 256, 16]
```

### Describe the problem
The documentation for tf.nn.convolution has the spatial dimensions of the output given as:

```
If padding == "SAME": output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides[i])
```

Which suggests that input_spatial_shape[0] should not affect output_spatial_shape[1], as is the case in the code block above.

This problem arises when using dilated convolutions as part of a larger model containing recurrent layers, in which one spatial dimension is left undefined to allow for unrolling the recurrent layers out during training along the undefined dimension.

This might be related to a [previously fixed problem with undefined batch sizes](https://github.com/tensorflow/tensorflow/issues/4742).

**Comment by shivaniag** [comment]
@jbms could you please take a look into this. I am able to reproduce the problem.

**Comment by jbms** [comment]
This is a limitation in the shape inference.  It may be difficult to fix without introducing a fair amount of additional complexity/code duplication, which is why it hasn't already been fixed.  The difficulty lies in the fact that tensorflow has special support for constant-valued tensors whose values are known at graph creation time, but it doesn't have support for partially-constant tensors, where some but not all values are known at graph creation time.  A contribution to fix this would be welcome, though.  As a workaround, you can use Tensor.set_shape to manually fix the shape.

**Comment by tfboyd** [comment]
@yzhwang   This is a very old issue and I am not asking you to resolve it.  With the new dilated convolution support would this be magically resolved.  My guess is no, you do not have to keep this assigned to yourself after you check.

**Comment by yzhwang** [comment]
@tfboyd I just tried with new dilated conv API and now it does show the correct shape inference.

```
input_tensor = tf.placeholder(tf.float32, (10, None, 256, 3))

dilated = tf.nn.conv2d(input_tensor,
                            tf.zeros((3, 1, 3, 16)),
                            strides=[1,1,1,1],
                            dilations=[1, 2, 1, 1],
                            padding='SAME')

print(dilated.get_shape())
```
I'm not sure how to fix the shape inference for convolution() though.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by sushreebarsa** [comment]
@haraldurt Instead of relying on "SAME" padding, we could try specifying the padding values manually to control the output shape more precisely. If you still have a concern then please create a new ticket as you are using an older version of TF which is not actively supported.
Thank you!
